### ISSUE TITLE
Decide on name for `Freeze`

### ISSUE THREAD

**Issue by traviscross** [rust-lang/rust#131401]
We still need to pick a name for `Freeze` (which may still be `Freeze`) so that we can proceed with:

- https://github.com/rust-lang/rfcs/pull/3633

Thoughts?

cc @rust-lang/lang

@rustbot labels +T-lang +I-lang-nominated

**Comment by p-avital** [comment]
I'm reiterating my proposal from Zulip: keeping `Freeze`, and instead giving `llvm`'s `freeze` a more explicit name. To quote myself from there:

> I might just be misunderstanding the docs for the intrinsic, but it feels like it'd translate well to a `MaybeUninit::read` (as opposed to `assume_init_read`), or maybe `deterministic_read` or something alike to avoid it looking simpler than the current `assume_init` set

**Comment by traviscross** [comment]
Let's see whether we have consensus to simply call it `Freeze`.

@rfcbot fcp merge

**Comment by joshtriplett** [comment]
> I'm reiterating my proposal from Zulip: keeping `Freeze`, and instead giving `llvm`'s `freeze` a more explicit name. To quote myself from there:
> 
> > I might just be misunderstanding the docs for the intrinsic, but it feels like it'd translate well to a `MaybeUninit::read` (as opposed to `assume_init_read`), or maybe `deterministic_read` or something alike to avoid it looking simpler than the current `assume_init` set

I don't know if `Freeze` is the best name for this trait, but I definitely agree that `freeze` is not the best name for the the LLVM concept.

**Comment by joshtriplett** [comment]
@rfcbot concern should-be-self-documenting-about-interior-mutability

I think we should use a name that's self-documenting and evokes the concept of interior mutability in some way.

**Comment by Scripter17** [comment]
Would `Frozen` work?

**Comment by Amanieu** [comment]
I'm partial to the name `InteriorImmutable` for the `Freeze` trait (which was actually one of the alternatives listed in the RFC). All of the cell-related documentation already uses the term "interior mutability" and the trait name makes it clear that it means that the type does not contain a cell.

**Comment by traviscross** [comment]
Stylistically, we don't end trait names with "-able".  We have `Copy` and not `Copyable`, e.g.  We just leaned on this principle in deciding to go with `derive(CoercePointee)` rather than various "-able" options in:

- https://github.com/rust-lang/rust/issues/129104

That is, trait names imply the capability to do the thing named in the trait.

What would be the version of an "interior immutable"-style name that matches that?

**Comment by Amanieu** [comment]
While "mutable" has a verb form "mutate", there isn't any verb equivalent of "immutable" ("not mutate"?). I would therefore argue for an exception to the -able policy for this specific case since "immutable" very clearly captures what this trait is promising.

An alternative that also captures the concept of interior mutability are might be `NoInteriorMutate`, but that just looks worse than `InteriorImmutable` since we're just arbitrarily avoiding the -able suffix.

**Comment by RalfJung** [comment]
The "action" this trait lets one do is know that if we do two reads from (pointers derived from) a reference where the pointee has this trait, we will get the same value. I don't know how to make that into a nice trait name...

**Comment by thynson** [comment]
`InteriorFreeze`?

**Comment by traviscross** [comment]
I could live with `FreezeInterior`.  It has a similar ring to it as what we did with `CoercePointee`.

**Comment by p-avital** [comment]
@traviscross I'd prefer `Freeze` to `FreezeInterior`, as I don't feel the addition of interior buys any clarity, while it does come at the cost of renaming which would make split resources for people looking it up; but at this point, I'm just ready to go with anything that grabs enough traction to finally get stabilized 🫠

**Comment by traviscross** [comment]
Similarly, I'd still prefer plain old `Freeze`.

**Comment by p-avital** [comment]
@traviscross This RFC is becoming stale again, with bike-shedding around the trait's name looking like the main blocker.

Do you think setting a deadline on the decision of naming (and possibly a poll with some of the existing options) could help push forward?

**Comment by traviscross** [comment]
It's a good point.  We just need to allocate some time for it in a meeting and commit to the bikeshed.  There's a meeting in mid December we have scheduled that would probably make sense for that.

**Comment by p-avital** [comment]
@traviscross Any timeline to try to get this sorted? The breaking change that pushed the need for this is still a major thorn in `stabby`'s side :(

**Comment by traviscross** [comment]
No, sorry.  What I can say is that clearing the backlog of things like this is high on my priority list, and I have some plans about how we might do that.

**Comment by hkBst** [comment]
Would it make sense to instead of the trait Freeze, have its negation NoFreeze, expressing the fact that some type does have an UnsafeCell somewhere in its representation, instead of the absence of that? Name could be InteriorMut(ate) or InternalChange or InnerMut(ate) or InnerChange or something like that...

"core::marker::Freeze is a trait that is implemented for any type whose memory layout doesn't contain any UnsafeCell: it indicates that the memory referenced by &T is guaranteed not to change while the reference is live." -- the RFC

**Comment by p-avital** [comment]
The problem with that is that since you can't make negative bounds, such a trait wouldn't have much use, as its usefulness stands from the particular properties `Freeze` allows you (or the compiler) to tell about a value.

On the other hand, I've wanted negative bounds for so long that if this bikesheding ends up being the final nail in that coffin, I'd wouldn't even be mad; but I somehow doubt that would happen :D

**Comment by traviscross** [comment]
> Is there any way I can be helpful in the process? :)

We may need a design meeting to resolve this, and for that we'll need a design document that collects the options that have been proposed and the arguments that have been made about them and then proposes a plan.  If you could do that, it'd help.  I'll reach out separately about this on Zulip.

**Comment by bstrie** [comment]
Given that the major caveat of this trait is that it only implies immutability up until some form of indirection, I'm surprised that nobody has proposed `Direct` as a component of a potential name. Something like `DirectFreeze` or `DirectlyImmutable` seem natural and accurate to me.

**Comment by hkBst** [comment]
I wanted to brain storm a little about what action this trait enables, so I went back to the RFC:

"Orthogonally to static-promotion, `core::marker::Freeze` can also be used to ensure that transmuting `&T` to a reference to an interior-immutable type (such as `[u8; core::mem::size_of::<T>()]`) is sound (as interior-mutation of a &T may lead to UB in code using the transmuted reference, as it expects that reference's pointee to never change). This is notably a safety requirement for zerocopy and bytemuck"

"static-promotion" seems to be about "previously, const REF: &T = &expr; was (accidentally) accepted even when expr may contain interior mutability. Now this requires that the type of expr satisfies T: core::marker::Freeze" so should perhaps be changed to "const-promotion" (with or without dash), so that gives the first option of ConstPromote.

Then there is the conversion from `&T` to `&[u8; core::mem::size_of::<T>()]`. Specifically in the presence of another `&T` and where `T` contains an UnsafeCell (not merely points to one), because then the other ref could be used to change the bit pattern of the memory of `T`, which I believe would be immediate UB in the presence of a converted `&[u8; core::mem::size_of::<T>()]`, because that shared ref implies constancy of the memory of `T` which would be violated by doing inner mutation. So that gives variations on TransmuteRefToBytesRef.

Then I wondered what the purpose is of this ref transmute. Is it for enabling some optimization? That would give OptimizeX. Bytemuck seems to use [Zeroable](https://docs.rs/bytemuck/latest/bytemuck/trait.Zeroable.html), [NoUninit](https://docs.rs/bytemuck/latest/bytemuck/trait.NoUninit.html), [AnyBitPattern](https://docs.rs/bytemuck/latest/bytemuck/trait.AnyBitPattern.html) among [other traits](https://docs.rs/bytemuck/latest/bytemuck/index.html#traits), while ZeroCopy has [KnownLayout](https://docs.rs/zerocopy/latest/zerocopy/trait.KnownLayout.html), [Immutable](https://docs.rs/zerocopy/latest/zerocopy/trait.Immutable.html), [Unaligned](https://docs.rs/zerocopy/latest/zerocopy/trait.Unaligned.html), and [IntoBytes](https://docs.rs/zerocopy/latest/zerocopy/trait.IntoBytes.html). Actually it seems not Immutable (as claimed by the RFC) but this IntoBytes is what is used for this transmuting meaning of Freeze: "Any IntoBytes type can be converted to a slice of initialized bytes of the same size. This is useful for efficiently serializing structured data as raw bytes." Why does the RFC (on a very cursory viewing) seem to suggest that Immutable and IntoBytes are the same? Or perhaps, why does zerocopy separate these traits?

Options:
- ConstPromote
- variations on transmuting references to T to references to bits:
  - TransmuteRefToBytesRef
  - ConvertRefToMemRef
  - ViewRefAsBitsRef
  - ViewRefAsRefToBits
  - ViewBorrowAsBorrowedBits
  - ViewBorrowAsRawMemoryBorrow
  - ViewAsShared
  - many other options
- OptimizeX ?

**Comment by traviscross** [comment]
We ended up deciding on `NoCell` elsewhere.

@rfcbot cancel
